site/content/blog/2019-10-16-debugging-using-charles.md

```diff
@@ -0,0 +1,93 @@
+* `static/*` are the JavaScript bundles, CSS, images, and plugin resources used the single page application
+
+Note that the server doesn't actually expect a request for `/index.html`, but serves up this file on any number of paths:
+
+* `/` landing on the site without any path
+* `/login` following a login link from an email, or just reloading the page
+* `/<team>/` any configured team, such as `core` or our internal `private-core` teams
+* `/<team>/<channel>` a channel on any configured team
+* `/<team>/pl/<post>` a permalink to a post
+* (endless other combinations, including unrecognized paths)
+
+Once the bootstrapping code is served, the application examines the current path and renders the appropriate part of application.
+
+In addition to the static assets, the server expects a number of programmatic interactions:
+
+* `/api/*` any REST API request
+* `/login/sso/saml` login flows via SAML
+* `/oauth/*` OAuth application flows
+* (various similar endpoints)
+
+Unfortunately, the overlap here between static assets and programmatic interactions makes splitting the routing slightly tricky when redirecting requests locally. For example:
+
+* `/login` should be redirected so as to serve the local `index.html`
+* `/login/sso/saml` should be sent to the server to avoid breaking SAML login
+
+## Overriding `community*.mattermost.com` static assets
+
+To configure Charles to redirect requests for static frontend assets, open the `Tools` menu and choose `Map Remote...`:
+
+![image](/blog/2019-10-16-debugging-using-charles/map-remote.png)
+
+(To avoid having to configure these manually, you can also just download and import [map-remote.xml](/blog/2019-10-16-debugging-using-charles/map-remote.xml).)
+
+The rules above yield the following results:
+
+* `/login` is sent to `localhost:8065/login`, serving up `index.html` during the login flow
+* `/static/*` is sent to `localhost:8065/static/*`, serving up all static assets
+* `/core/*` is sent to `localhost:8065/core/*`, serving up `index.html` for any link on the core team
+* `/private-core/*` is sent to `localhost:8065/core/*`, serving up `index.html` for any link on the private core team
+* all other paths continue to route directly to the community server
+
+To start your debugging session, first make sure your local Mattermost server is running by following the [Developer Setup](https://developers.mattermost.com/contribute/webapp/developer-setup/) instructions. Then, browse to `/login` against the community server of your choice. If you're not already logged in, this will allow you to complete any necessary login flow (even through SAML!), but you'll get sent back to `/` and what appears to just be a blank page with console errors. Manually head back to `/login` and you'll find yourself successfully logged in and on community with all static assets being served up by your local development server instance:
```

Comment:
```suggestion
To start your debugging session, first make sure your local Mattermost server is running by following the [Developer Setup](https://developers.mattermost.com/contribute/webapp/developer-setup/) instructions. Then, browse to `/login` against the community server of your choice. If you're not already logged in, this will allow you to complete any necessary login flow (even through SAML!), but you'll get sent back to `/` and what appears to just be a blank page with console errors. Manually head back to `/login` and you'll find yourself successfully logged in on community with all static assets being served up by your local development server instance:
```